gio/src/cancellable_future.rs

```diff
@@ -0,0 +1,143 @@
+// Take a look at the license at the top of the repository in the LICENSE file.
+
+use crate::cancellable::CancellableExtManual;
+use crate::prelude::CancellableExt;
+use crate::Cancellable;
+use crate::IOErrorEnum;
+use pin_project_lite::pin_project;
+use std::future::Future;
+use std::pin::Pin;
+use std::task::Context;
+use std::task::Poll;
+
+pin_project! {
+    /// A future which can be cancelled via [`Cancellable`].
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// let c = Cancellable::new();
+    /// ctx.spawn_local(CancellableFuture::new(task, c)
+    ///                            .await
+    ///                            .map(|_| ()))
+    ///
+    /// ```
+    pub struct CancellableFuture<F> {
+        #[pin]
+        future: F,
+        cancellable: Cancellable,
+    }
+}
+
+impl<F> CancellableFuture<F> {
+    /// Creates a new `CancellableFuture` using a [`Cancellable`].
+    ///
+    /// When [`cancel`](CancellableExt::cancel) is called, the future will complete
+    /// immediately without making any further progress. In such a case, an error
+    /// will be returned by this future (i.e., [`IOErrorEnum::Cancelled`]).
+    pub fn new(future: F, cancellable: Cancellable) -> Self {
+        Self {
+            future,
+            cancellable,
+        }
+    }
+
+    /// Checks whether the future has been cancelled.
+    ///
+    /// Note that all this method indicates is whether [`cancel`](CancellableExt::cancel)
+    /// was called. This means that it will return true even if:
+    ///   * `cancel` was called after the future had completed.
+    ///   * `cancel` was called while the future was being polled.
+    #[inline]
+    pub fn is_cancelled(&self) -> bool {
+        self.cancellable.is_cancelled()
+    }
+}
+
+impl<F> Future for CancellableFuture<F>
+where
+    F: Future,
+{
+    type Output = Result<<F as Future>::Output, glib::Error>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        if self.is_cancelled() {
+            return Poll::Ready(Err(glib::Error::new(
+                IOErrorEnum::Cancelled,
+                "Task cancelled",
+            )));
+        }
+
+        match self.as_mut().project().future.poll(cx) {
+            Poll::Ready(out) => Poll::Ready(Ok(out)),
+            Poll::Pending => {
+                self.cancellable.connect_cancelled({
```

Comment:
This should probably only happen once. It's absolutely valid to call `poll()` over and over again after `Pending` is returned without waiting for the waker.

Also I think if the waker changes (and it can change), you will want to disconnect for the older waker

Reply:
Yeah, the waker can change and it should only use the most recent one.

> Note that on multiple calls to poll, only the [Waker](https://doc.rust-lang.org/std/task/struct.Waker.html) from the [Context](https://doc.rust-lang.org/std/task/struct.Context.html) passed to the most recent call should be scheduled to receive a wakeup.

https://doc.rust-lang.org/std/future/trait.Future.html#tymethod.poll

Reply:
Done. Thanks for the catch